workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/TextController.kt

```diff
@@ -0,0 +1,166 @@
+   * is not provided at subscription time. Workflows can safely observe changes by
+   * converting this value to a worker. (When using multiple instances, remember to provide unique
+   * key values to each `asWorker` call.)
+   *
+   * If you can do processing that doesn't require running a `WorkflowAction` or triggering a render
+   * pass, it can be done in regular Flow operators before converting to a worker.
+   */
+  public val onTextChanged: Flow<String> =
+  // Drop the first value because the StateFlow will always immediately emit the current value on
+    // subscription.
+    _textValue.drop(1)
+
+  /**
+   * The current text value.
+   */
+  public var textValue: String
+    get() = _textValue.value
+    set(value) {
+      _textValue.value = value
+    }
+
+  /**
+   * Call this from your LayoutRunner's showRendering method. This method is idempotent: if it has
+   * already been called with a particular [EditText], and the view has not been detached since the
+   * last call, it will do nothing. If a different [TextController]'s [control] is called on the same
+   * [EditText], the old one will be disconnected and the new one will replace it.
+   *
+   * See [TextController] for more documentation.
+   */
+  public fun control(view: EditText) {
+    // Do nothing if already subscribed on a previous update pass and the coroutine is still active.
+    val registeredController =
+      view.getTag(R.id.text_controller_rendering) as? TextControllerSubscription
+    if (registeredController?.controller === this &&
+      // This check ensures the subscription is re-started if the view was somehow detached since the
+      // call, eg. in a RecyclerView.
+      registeredController.subscription.isActive
+    ) {
+      return
+    }
+
+    // If we're replacing a controller with a different one, cancel the previous subscription.
+    registeredController?.subscription?.cancel()
+
+    view.setText(_textValue.value)
+    val subscription = view.launchWhenAttached {
+      _textValue
+        .onEach { textValue ->
+          // Only set the text if the actual text content has changed.
+          if (textValue != view.text.toString()) {
+            view.setText(textValue)
+          }
+        }
+        .launchIn(this)
+
+      view.listenForTextChangesUntilCancelled {
+        _textValue.value = it?.toString().orEmpty()
+      }
+    }
+    view.setTag(R.id.text_controller_rendering, TextControllerSubscription(this, subscription))
+  }
+}
+
+@OptIn(WorkflowUiExperimentalApi::class)
+private class TextControllerSubscription(
+  val controller: TextController,
+  val subscription: Job
+)
+
+/**
+ * Suspends the coroutine until cancelled, calling [handler] any time a text change event is fired.
+ */
+private suspend fun TextView.listenForTextChangesUntilCancelled(
+  handler: (CharSequence?) -> Unit
+): Nothing {
+  suspendCancellableCoroutine<Nothing> { continuation ->
```

Comment:
I know I've reviewed before but what a brilliant use of structured concurrency. This is a great advertisement for the power of coroutines.

Reply:
Couldn't agree more. All kudos to @zach-klippenstein, I'm just a stenographer here.

Reply:
In fact, I should fix the attribution on that commit.